default/python/AI/Behavior/Character.py

```diff
@@ -0,0 +1,517 @@
+        """Deprecated"""
+        return None
+
+    def may_research_xeno_genetics_variances(self):  # pylint: disable=no-self-use,unused-argument
+        """Return True if able to vary xeno genetics research"""
+        return True
+
+    def prefer_research_defensive(self):  # pylint: disable=no-self-use,unused-argument
+        """Return True if should prefer defensive tech research"""
+        return True
+
+    def prefer_research_low_aggression(self):  # pylint: disable=no-self-use,unused-argument
+        """Return True if should prefer less aggressive tech research"""
+        return True
+
+    def may_research_tech(self, tech):  # pylint: disable=no-self-use,unused-argument
+        """Return True if permitted to research ''tech''"""
+        return True
+
+    def may_research_tech_classic(self, tech):  # pylint: disable=no-self-use,unused-argument
+        """Return True if permitted to research ''tech'' in the classic algorithm"""
+        return True
+
+    def attitude_to_empire(self, other_empire_id, diplomatic_logs):  # pylint: disable=no-self-use,unused-argument
+        """Return a value from [-10 .. 10] representing attitude towards other empire."""
+        return None
+
+
+class Aggression(Behavior):
+    """A behavior that models level of difficulty and aggression."""
+    # TODO copy all aggression decisions into this class
+    # TODO break this class into two behaviors: level of difficulty and aggression
+
+    def __init__(self, aggression):
+        self.aggression = aggression
+
+    @property
+    def key(self):
+        return self.aggression
+
+    def may_explore_system(self, monster_threat):
+        return monster_threat < 2000 * self.aggression
+
+    def may_surge_industry(self, total_pp, total_rp):
+        return ((self.aggression > fo.aggression.cautious) and
+                ((total_pp + 1.6 * total_rp) < (50 * self.aggression)))
+
+    def may_maximize_research(self):
+        return self.aggression >= fo.aggression.maniacal
+
+    def max_number_colonies(self):
+        # significant growth barrier for low aggression, negligible for high aggression
+        return 2 + ((0.5+self.aggression)**2)*fo.currentTurn()/50.0
+
+    def may_invade(self):
+        return (self.aggression > fo.aggression.turtle
+                and not (self.aggression == fo.aggression.beginner and fo.currentTurn() < 150))
+
+    def may_invade_with_bases(self):
+        return self.aggression > fo.aggression.typical
+
+    def invasion_priority_scaling(self):
+        return 0.5 if self.aggression == fo.aggression.beginner else 1.0
+
+    def military_priority_scaling(self):
+        return 1.0 if self.aggression > fo.aggression.typical else ((1.0 + self.aggression) / (1.0 + fo.aggression.typical))
+
+    def max_defense_portion(self):
+        return [0.7, 0.4, 0.3, 0.2, 0.1, 0.0][self.aggression]
+
+    def check_orbital_production(self):
+        aggression_index = max(1, self.aggression)
```

Comment:
arguably deserves its own function as it is used more than once and is non-trivial.